### PR TITLE
Added zones in exact order (still testing)

### PR DESCRIPTION
**Pull request by Thyreal**
Hi, 
I added a Function to only trigger a Message wenn the zones are entered in an exact order. Use case is when someone comes up to the door i get a message, wenn someone is leaving i dont. 
This is my first gitup pull request so forgive me if i overlook on something. 
I am still in the testing but i think it is working.

### CODE REVIEW

**Comment by SgtBatten** [comment]
Thanks for doing this. I look forward to testing. 

I note that there are significant number of changes recorded but it's mostly due to formatting issues. In order to merge these I'll need you to start with the version currently on this repository and make only the changes necessary for the zone order to work, then submit that. 

A simple way to do it is to open the file on this repository and start editing it, when you go to save it it will be able to be turned into a pull request.

File: Frigate Camera Notifications/Beta [hunk 1]
```diff
@@ -1138,7 +1151,10 @@ action:
                     last_zones: "{{ event['before']['entered_zones']|lower}}"
                     enteredzones: "{{ event['after']['entered_zones']|lower}}"
                     zone_filter: "{{ not zone_only or zones|select('in', enteredzones)|list|length > 0 }}"
-                    zone_multi_filter: "{{not zone_only or not zone_multi or ( enteredzones|list|length > 0 and zones and zones|reject('in', enteredzones)|list|length == 0 ) }}"
+                    zone_multi_filter: "{{not zone_only and zone_multi and not zone_multi_ex and ( enteredzones|list|length > 0 and zones and zones|reject('in', enteredzones)|list|length == 0 ) }}"
+                    zone_multi_ex_filter: >
```
Comment (SgtBatten) [inline Frigate Camera Notifications/Beta:1155]:
```suggestion
                    zone_multi_filter: "{{not zone_only and zone_multi and not zone_multi_ex and ( enteredzones|list|length > 0 and zones and zones|reject('in', enteredzones)|list|length == 0 ) }}"
                    zone_multi_exact_order_filter: >
```

File: Frigate Camera Notifications/Beta [hunk 2]
```diff
@@ -1218,7 +1234,9 @@ action:
                                       Multi Zone Enabled: {{zone_multi}}, 
                                       Required Zones: {{input_zones}}, 
                                       Last Zones: {{last_zones}}, 
-                                      Entered zones: {{enteredzones|list|length}} - {{enteredzones}}, 
+                                      Entered zones: {{enteredzones|list|length}} - {{enteredzones}},
+                                      Multi Zone Filter: {{zone_multi_filter}},
+                                      Multi Zone Filter Exact: {{zone_multi_ex_filter}},
                                       TEST: {{'PASS' if not zone_only or (not zone_multi and zones|select('in', enteredzones)|list|length ) else 'PASS (Multi)' if zone_only and zone_multi and zone_multi_filter else 'FAIL (Multi)' if zone_only and zone_multi else 'FAIL' }},
```
Comment (SgtBatten) [inline Frigate Camera Notifications/Beta:1240]:
```suggestion
                                      Entered zones: {{enteredzones|list|length}} - {{enteredzones}},
                                      Multi Zone Filter Exact: {{zone_multi_exact_order_filter}},
                                      TEST: {{'PASS' if not zone_only or (not zone_multi and not zone_multi_exact_order and zones|select('in', enteredzones)|list|length ) else 'PASS (Multi)' if zone_only and zone_multi and zone_multi_filter and not zone_multi_exact_order else 'FAIL (Multi)' if zone_only and zone_multi and not zone_multi_exact_order else 'PASS (Exact Order)' if zone_multi_exact_order and zone_multi_exact_order_filter else 'FAIL (Exact Order)' if zone_multi_exact_order else 'FAIL'}}, 
```

File: Frigate Camera Notifications/Beta [hunk 3]
```diff
@@ -1235,15 +1253,28 @@ action:
...
+                                    entered_zones_changed:{{entered_zones_changed}},
+                                    sub_label_changed:{{sub_label_changed}},
+                                    Test Condition loop: "{{ loitering or (custom_filter and not home and zone_filter and (zone_multi_filter or zone_multi_ex_filter) and state_true and (new_snapshot or presence_changed or stationary_moved or zone_only_changed or entered_zones_changed or sub_label_changed)) }}",
                 - alias: "Custom Action Auto Multi"
                   choose:
                     - conditions:
                         - "{{ custom_action_auto_multi | length > 0 }}"
-                        - "{{ loitering or (not home and zone_filter and zone_multi_filter and state_true and (new_snapshot or presence_changed or stationary_moved or zone_only_changed or entered_zones_changed or sub_label_changed)) }}"
+                        - "{{ loitering or (not home and zone_filter and (zone_multi_filter or zone_multi_ex_filter) and state_true and (new_snapshot or presence_changed or stationary_moved or zone_only_changed or entered_zones_changed or sub_label_changed)) }}"
```
Comment (SgtBatten) [inline Frigate Camera Notifications/Beta:1273]:
```suggestion
                        - "{{ loitering or (not home and zone_filter and (zone_multi_filter or zone_multi_exact_order_filter) and state_true and (new_snapshot or presence_changed or stationary_moved or zone_only_changed or entered_zones_changed or sub_label_changed)) }}"
```

File: Frigate Camera Notifications/Beta [hunk 4]
```diff
@@ -1235,15 +1253,28 @@ action:
...
                       sequence: !input "custom_action_auto_multi"
                 - alias: "Update Notification"
                   choose:
-                    - conditions: "{{ loitering or (custom_filter and not home and zone_filter and zone_multi_filter and state_true and (new_snapshot or presence_changed or stationary_moved or zone_only_changed or entered_zones_changed or sub_label_changed)) }}"
+                    - conditions: "{{ loitering or (custom_filter and not home and zone_filter and (zone_multi_filter or zone_multi_ex_filter) and state_true and (new_snapshot or presence_changed or stationary_moved or zone_only_changed or entered_zones_changed or sub_label_changed)) }}"
```
Comment (SgtBatten) [inline Frigate Camera Notifications/Beta:1277]:
```suggestion
                    - conditions: "{{ loitering or (custom_filter and not home and zone_filter and (zone_multi_filter or zone_multi_exact_order_filter) and state_true and (new_snapshot or presence_changed or stationary_moved or zone_only_changed or entered_zones_changed or sub_label_changed)) }}"
```

File: Frigate Camera Notifications/Beta [hunk 5]
```diff
@@ -1235,15 +1253,28 @@ action:
...
+                                    zone_multi_filter:{{zone_multi_filter}},
+                                    zone_multi_ex_filter:{{zone_multi_ex_filter}},
+                                    state_true: {{state_true}},
+                                    new_snapshot:{{new_snapshot}},
+                                    presence_changed: {{presence_changed}},
+                                    stationary_moved: {{stationary_moved}},
+                                    zone_only_changed: {{zone_only_changed}},
+                                    entered_zones_changed:{{entered_zones_changed}},
+                                    sub_label_changed:{{sub_label_changed}},
+                                    Test Condition loop: "{{ loitering or (custom_filter and not home and zone_filter and (zone_multi_filter or zone_multi_ex_filter) and state_true and (new_snapshot or presence_changed or stationary_moved or zone_only_changed or entered_zones_changed or sub_label_changed)) }}",
```
Comment (SgtBatten) [inline Frigate Camera Notifications/Beta:1268]: leave for now but probably remove this once testing is done.

**Review by SgtBatten — CHANGES_REQUESTED** [review]

File: Frigate Camera Notifications/Beta (hunk 5 above)
Comment (Thyreal) [reply Frigate Camera Notifications/Beta:1267]: yeap that was the plan ;-)

File: Frigate Camera Notifications/Beta [hunk 6]
```diff
@@ -1138,7 +1151,10 @@ action:
                     last_zones: "{{ event['before']['entered_zones']|lower}}"
                     enteredzones: "{{ event['after']['entered_zones']|lower}}"
                     zone_filter: "{{ not zone_only or zones|select('in', enteredzones)|list|length > 0 }}"
-                    zone_multi_filter: "{{not zone_only or not zone_multi or ( enteredzones|list|length > 0 and zones and zones|reject('in', enteredzones)|list|length == 0 ) }}"
+                    zone_multi_filter: "{{not zone_only and zone_multi and not zone_multi_ex and ( enteredzones|list|length > 0 and zones and zones|reject('in', enteredzones)|list|length == 0 ) }}"
```
Comment (SgtBatten) [inline Frigate Camera Notifications/Beta:1154]:
```suggestion
                    zone_multi_filter: "{{not zone_only and zone_multi and ( enteredzones|list|length > 0 and zones and zones|reject('in', enteredzones)|list|length == 0 ) }}"
```

File: Frigate Camera Notifications/Beta [hunk 7]
```diff
@@ -1138,7 +1151,10 @@ action:
...
+                    zone_multi_exact_order_filter: >
+                      {% set result = (zones == enteredzones[:zones|length]) %}
+                        {{ zone_only and zone_multi_ex and result and not zone_multi }}
```
Comment (SgtBatten) [inline Frigate Camera Notifications/Beta:1157]:
```suggestion
                        {{ zone_only and zone_multi_exact_order and result}}
```

**Comment by JonGilmore** [comment]
Just wanted to drop in to express my support for this, it'd be awesome to have this. I currently do this with a [yaml automation (with telegram alerts)](https://github.com/JonGilmore/ha-personal/blob/e0f5d981a860830248e9851e14aaf89a340a22e9/automation/frigate.yaml#L162). I'm happy to help test things out, let me know if you need further help. I've linked my automation if you want to see how I'm doing it in yaml.

**Comment by SgtBatten** [comment]
Yes it looks ot be possible. Unfortunately after asking @Thyreal to switch to the beta branch that's now focusing on frigate 0.14 so we actually need to move these changes back onto stable.

**Pull request closed without merging**